### DISCUSSION TITLE
Websocket api not showing in auto-generated documentation

### DISCUSSION THREAD

**Discussion by javierjsa** [fastapi/fastapi#7264] [Questions] [answered]
Hi,

I've implemented a small api with bot an http and a websocket interface. The http inteface shows up in the auto-generated documentation, but the websocket interface does not. I suspect this feature is missing, as **@router.websocket** does not have a _tags_  param, however I'm not sure. Are there any options apart from linking to an external source?

Regards

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [ ] I searched the FastAPI documentation, with the integrated search. -> stuck on intializing search, searched manually
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

**Comment by zero-shubham** [comment]
does OpenAPI 3 has any spec for Websocket / MQTT ? There's something new for webhooks. People suggest to use asyncapi.com specs for documenting such message driven APIs so may be an opt-in option for that?

**Comment by flokasper** [comment]
Same problem here.
OpenAPI 3 doesn't have anything for websocket, so the generated json file, Swagger UI and Redoc can't be used
The closest specification is asyncapi. A package like [fastapi.openapi](https://github.com/tiangolo/fastapi/tree/master/fastapi/openapi) have to be written to generate the json file. Also I didnt find any js template that generate a doc dynamically, I only found CLI. 
It would be a nice feature

**Comment by javierjsa** [comment]
So this would be a candidate for feature request, then

**Comment by tiangolo** [comment, marked as answer]
Thanks for the help here everyone! :clap: :bow: 

Indeed, the automatic docs UI is done by Swagger UI, which reads and integrates with OpenAPI.

WebSockets is still quite new, and there are not many standards for it. They mainly provide a two-way communication channel, so they are closer to TCP than to HTTP, and so, it's not very clear what a standard for communicating WebSocket data would look like, or how it would be documented, as they give you all the "freedom" without any of the restrictions. And those restrictions are what a more "formal" API standard like OpenAPI is based on.

So, in terms of showing them in the docs, it would be a feature request for Swagger UI, not for FastAPI. But that would require first for them to be definable in OpenAPI, in the spec. But that would require them to be "definable", so it would more or less require a consensus about how to use them, apart from a completely free two-way communication channel.
